Align Vue to other modules (ESLint + TSConfig)

charles-allen opened the pull request:
## ✅ DoD

- [x] All work is complete
- [x] Migrations: NA
- [x] Documentation: NA
- [x] Release notes: NA
- [x] Issues linked: _none_

## 🛑 Problems

- If we upgrade to TypeScript 4.3 (currently beta), `ts-essentials` seems to break. Hopefully they will fix it before then.

## 📝 Summary

- All modules now use the same `.eslintrc`, `.prettierrc`, `tsconfig.json`
  - Separated ESLint config for JS; extend JS config to create TS & Vue configs
  - Merge configs together based on file patterns to create a universal config
  - Move `.prettierrc` & `.stylelintrc` to root
- Support for linting only the root directory (`yarn lint-root`/`yarn fix-root`)
- Fix some lint errors (including root configs)

charles-allen committed:
[Lint] Support `lint` on only root directory

charles-allen committed:
[Lint] Decouple JS/TS; Vue now extends JS

charles-allen committed:
[Lint] Move .prettierrc/.stylelintrc to root

charles-allen committed:
[Lint] Auto-fix for new rules

charles-allen committed:
[Lint] Fix easy errors

charles-allen committed:
[TypeScript] Stricter config; Vue extends base

charles-allen committed:
[Lint] Fix broken rebuild

antonyharfield commented on .eslintrc-js.js:
```diff
@@ -0,0 +1,153 @@
+    'woke/LGBTQ': 'error',
+    'woke/gender': 'error',
+    'woke/profanity': 'error',
+    'woke/racism': 'off' // need to find another abbreviation for "production"
```
This could be interpreted badly... "we turned our racism filters off"!

charles-allen replied:
Technically this is off by default on all projects; I just haven't turned it on! Though it does look worse that it's explicitly off instead of implicitly.

To fix this we need to update our vocabulary. We're currently in violation because we use the word `prod` (which I discovered is an offensive term).

charles-allen replied:
This is now fixed! 

I tracked down the offensive abbreviations and replaced them with `production` (which also has the added effect of making it slightly harder to mess up production).

antonyharfield reviewed (approved):
For new projects, I need an option on yarn for: `yarn init --copying-chaz-linting-defaults`

charles-allen commented:
> For new projects, I need an option on yarn for: `yarn init --copying-chaz-linting-defaults`

I think this probably exists, provided I publish a config. I'm considering it!

charles-allen committed:
[Docs] Clean up quick-start

grindarius reviewed (approved):
please provide `eslint-config-charles-allen` on npm, thank you. this PR looks good to me.

charles-allen committed:
[Lint] Fix woke rules

charles-allen committed:
[Lint] Auto-format hosting configs

charles-allen committed:
[Lint] Fix `unicorn/no-array-callback-reference`

charles-allen committed:
[Lint] Fix `unicorn/no-nulls`

charles-allen committed:
[Lint] Fix `unicorn/no-keyword-prefix`

charles-allen committed:
[Lint] Remove lodash rules

charles-allen committed:
[Lint] Move shared devDeps to root

- Remove unused eslint-plugin-lodash devDep
- Add lodash to hosting deps
- Restore experimentalDecorators flag

charles-allen committed:
[Deps] Centralize devDeps

- Move types, TypeScript, lint, testing to root
- Add all missing peer deps
- Bump everything (except locked)

charles-allen committed:
[Lint] Prettier woke up...

charles-allen committed:
[Lint] Sort prettier config

charles-allen committed:
Merge remote-tracking branch 'origin/develop' into feature/eslint-align

charles-allen committed:
[Lint] Remove last exception

charles-allen committed:
[Lint] Single-line for JS config :)

charles-allen committed:
[Lint] Manual single-line JS; Lint non-config JS

charles-allen committed:
[TypeScript] Clean up include/exclude

charles-allen committed:
[TypeScript] Fix return on all code paths
